File: src/main/java/edu/practice/datastructures/Heap.java
package edu.practice.datastructures;

import java.util.Arrays;

public class Heap<E extends Comparable<E>> {

    private static final int DEFAULT_INITIAL_CAPACITY = 11;

    E[] heap;
    private final boolean isMinHeap;
    private int size;

    public Heap() {
        this(true);
    }

    public Heap(boolean isMinHeap) {
        this(DEFAULT_INITIAL_CAPACITY, isMinHeap);
    }

    @SuppressWarnings("unchecked")
    public Heap(int initialCapacity, boolean isMinHeap) {
        if (initialCapacity < 1) {
            throw new IllegalArgumentException();
        }

        this.heap = (E[]) new Comparable[initialCapacity];
        this.isMinHeap = isMinHeap;
        size = 0;
    }

    private boolean compare(E a, E b) {
        return isMinHeap ? (a.compareTo(b) < 0) : (a.compareTo(b) > 0);
    }

    private void swap(int i, int j) {
        E temp = heap[i];
        heap[i] = heap[j];
        heap[j] = temp;
    }

    private void heapifyUp(int index) {
        while (index > 0) {
            int parentIndex = (index - 1) / 2;

            if (compare(heap[index], heap[parentIndex])) {
                swap(index, parentIndex);
                index = parentIndex;
            } else {
                break;
            }
        }
    }

    private void heapifyDown(int index) {
        int leftChild = 2 * index + 1;
        int rightChild = 2 * index + 2;
        int smallestOrLargest = index;

        if (leftChild < size && compare(heap[leftChild], heap[smallestOrLargest])) {
            smallestOrLargest = leftChild;
        }

        if (rightChild < size && compare(heap[rightChild], heap[smallestOrLargest])) {
            smallestOrLargest = leftChild;
        }

        if (smallestOrLargest != index) {
            swap(index, smallestOrLargest);
            heapifyDown(smallestOrLargest);
        }
    }

    public void insert(E data) {
        if (size == heap.length) {
            // Resize needed
            heap = Arrays.copyOf(heap, heap.length * 2);
        }

        heap[size] = data;
        heapifyUp(size);
        size++;
    }

    public E remove() {
        if (isEmpty()) {
            throw new IllegalStateException("Heap is Empty.");
        }

        E root = heap[0];
        heap[0] = heap[size - 1];
        size--;

        heapifyDown(0);
        return root;
    }

    public E peek() {
        if (isEmpty()) {
            throw new IllegalStateException("Heap is Empty.");
        }

        return heap[0];
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return (size == 0);
    }
}

File: src/main/kotlin/datastructures/Heap.kt
package edu.practice.datastructures

class Heap<E: Comparable<E>>(
    initialCapacity: Int = DEFAULT_INITIAL_CAPACITY,
    val isMinHeap: Boolean = true
) {

    companion object {
        private const val DEFAULT_INITIAL_CAPACITY = 11
    }

    private var heap: ArrayList<E>
    private var size = 0

    init {
        if (initialCapacity < 1) {
            throw IllegalArgumentException("Initial Capacity of Heap should be at least 1")
        }

        heap = ArrayList(DEFAULT_INITIAL_CAPACITY)
    }

    private fun heapifyUp(startIndex: Int) {
        var index = startIndex

        while (index > 0) {
            val parentIndex = (index - 1) / 2

            if (!compare(heap[index], heap[parentIndex])) {
                break
            }

            swap(index, parentIndex)
            index = parentIndex
        }
    }

    private fun heapifyDown(startIndex: Int) {
        val leftChild = 2 * startIndex + 1
        val rightChild = 2 * startIndex + 2
        var smallestOrLargest = startIndex

        if (leftChild < size && compare(heap[leftChild], heap[smallestOrLargest])) {
            smallestOrLargest = leftChild
        }
        if (rightChild < size  && compare(heap[rightChild], heap[smallestOrLargest])) {
            smallestOrLargest = rightChild
        }
        if (smallestOrLargest != startIndex) {
            swap(startIndex, smallestOrLargest)
            heapifyDown(smallestOrLargest)
        }
    }

    fun insert(data: E) {
        heap.addLast(data)
        heapifyUp(size)
        size++
    }

    fun remove(): E {
        val item: E?
        try {
            item = peek()
        } catch (exception: IllegalStateException) {
            throw IllegalStateException("Cannot remove from heap when size of Heap is 0")
        }

        heap[0] = heap[size - 1]
        heapifyDown(0)

        size--
        return item
    }

    private fun swap(i: Int, j: Int) {
        val temp = heap[i]
        heap[i] = heap[j]
        heap[j] = temp
    }

    private fun compare(a: E, b: E): Boolean {
        return if (isMinHeap) a < b else a > b // Overloaded operator by Comparable
    }

    fun peek(): E {
        if (!isEmpty()) {
            return heap[0]
        } else {
            throw IllegalStateException("Cannot peek heap when size of Heap is 0")
        }
    }

    fun size(): Int = size

    fun isEmpty(): Boolean = (size == 0)
}